Add Dockerfile instructions to reset properties inherited from parent image

dqminh opened the pull request:
This is an attemp to fix https://github.com/docker/docker/issues/3465. I've been working on this yesterday and figure that the design can be reviewed at this stage.

New added instructions are:
- `UNSETENV ENV1 ENV2 ...` unset environment variables inherited from parent image
- `UNEXPOSE PORT ...`  remove exposed ports inherited from parent image
- `NOVOLUME PATH` remove volume mountpoints from parent image

dqminh commented:
@erikh @icecrime i rebased and changed the instructions. PTAL

Also Happy New Year :fireworks:

SvenDowideit commented:
Docs LGTM - @fredlf @jamtur01 

@dqminh can you perhaps describe an example which needs it?

dqminh commented:
@SvenDowideit i embedded one short, albeit contrived example for each instruction in `builder.md` file. Please let me know if those examples make sense to you.

SvenDowideit commented:
yes, very nice - your examples tell the story well :)

Docs LGTM - @fredlf @jamtur01

jamtur01 commented on docs/man/Dockerfile.5.md:
```diff
@@ -117,6 +117,11 @@ or
 
+**UNEXPOSE**
+ --**UNEXPOSE <port> [<port>...]**
+ The **UNEXPOSE** instruction instruction informs Docker that the container no
```
instruction twice.

jamtur01 commented on docs/man/Dockerfile.5.md:
```diff
@@ -130,6 +135,12 @@ or
 
+**UNSETENV**
+ --**UNSETENV <key> [<key>...]**
+ The **UNSETENV** instructions removed the specified environment variables from
```
instruction removes

jamtur01 commented on docs/man/Dockerfile.5.md:
```diff
@@ -179,6 +190,11 @@ or
 
+**NOVOLUME**
+ --**NOVOLUME ["/data"]**
+ The **NOVOLUME** informs Docker that the image no longer want to
```
The **NOVOLUME** instruction informs Docker that the images no longer wants to...

jamtur01 commented on docs/sources/reference/builder.md:
```diff
@@ -337,6 +340,27 @@ accessible from the host by default. To expose ports to the host, at runtime,
+to listen on the specified network ports at runtime.
+
+For example, given a image named `debug-app` defined with the following
+Dockerfile:
```
\`Dockerfile\`

jamtur01 commented on docs/sources/reference/builder.md:
```diff
@@ -337,6 +340,27 @@ accessible from the host by default. To expose ports to the host, at runtime,
+    FROM debian:wheezy
+    EXPOSE 3000 8080
+
+we want to create a new image which exposes only port 3000. The
```
We

jamtur01 commented on docs/sources/reference/builder.md:
```diff
@@ -337,6 +340,27 @@ accessible from the host by default. To expose ports to the host, at runtime,
+    EXPOSE 3000 8080
+
+we want to create a new image which exposes only port 3000. The
+Dockerfile for new image can be written as:
```
`\Dockerfile\`

jamtur01 commented on docs/sources/reference/builder.md:
```diff
@@ -379,6 +403,28 @@ change them using `docker run --env <key>=<value>`.
+
+    UNSETENV <key> [<key>...]
+
+The `UNSETENV` instructions removed the specified environment variables from
```
removes

jamtur01 commented on docs/sources/reference/builder.md:
```diff
@@ -379,6 +403,28 @@ change them using `docker run --env <key>=<value>`.
+instructions.
+
+For example, given a image named `debug-app` defined with the following
+Dockerfile:
```
\`Dockerfile\`

jamtur01 commented on docs/sources/reference/builder.md:
```diff
@@ -379,6 +403,28 @@ change them using `docker run --env <key>=<value>`.
+    FROM debian:wheezy
+    ENV DEBUG true
+
+we want to create a new image without the `DEBUG` environment variable. The
```
We

jamtur01 commented on docs/sources/reference/builder.md:
```diff
@@ -379,6 +403,28 @@ change them using `docker run --env <key>=<value>`.
+    ENV DEBUG true
+
+we want to create a new image without the `DEBUG` environment variable. The
+Dockerfile for new image can be written as:
```
\`Dockerfile\`

jamtur01 commented on docs/sources/reference/builder.md:
```diff
@@ -774,6 +820,33 @@ documentation.
+
+    NOVOLUME ["/data"]
+
+The `NOVOLUME` instruction informs Docker that the image no longer want to
```
wants

jamtur01 commented on docs/sources/reference/builder.md:
```diff
@@ -774,6 +820,33 @@ documentation.
+> you must use double-quotes (") around words not single-quotes (').
+
+For example, given a image named `debug-app` defined with the following
+Dockerfile:
```
\`Dockerfile\`

jamtur01 commented on docs/sources/reference/builder.md:
```diff
@@ -774,6 +820,33 @@ documentation.
+    VOLUME ["/data", "/var/log"]
+
+we want to create a new image without `/var/log` volume. The
+Dockerfile for new image can be written as:
```
\`Dockerfile\`

jamtur01 commented on docs/sources/reference/builder.md:
```diff
@@ -774,6 +820,33 @@ documentation.
+    FROM debian:wheezy
+    VOLUME ["/data", "/var/log"]
+
+we want to create a new image without `/var/log` volume. The
```
We

jamtur01 commented:
Docs needs some work. Also update the man pages?

icecrime commented:
Ping @tiborvass for review.

dqminh commented:
@jamtur01 i rebased and updated the docs. PTAL. And the `man` pages you mentioned is `Dockerfile.5.md` right, which already has the new instructions' docs ?

jamtur01 commented:
Docs LGTM Ping @SvenDowideit @fredlf

SvenDowideit commented:
Docs LGTM - @fredlf

fredlf commented on docs/sources/reference/builder.md:
```diff
@@ -343,6 +346,27 @@ accessible from the host by default. To expose ports to the host, at runtime,
+The `UNEXPOSE` instruction informs Docker that the container no longer wants
+to listen on the specified network ports at runtime.
+
+For example, given a image named `debug-app` defined with the following
```
an image

fredlf commented on docs/man/Dockerfile.5.md:
```diff
@@ -117,6 +117,11 @@ or
+**UNEXPOSE**
+ --**UNEXPOSE <port> [<port>...]**
+ The **UNEXPOSE** instruction informs Docker that the container no
+ longer wants to listen on the specified network ports at runtime.
```
Given that containers aren't sentient (yet), they can't "want" things. :-) Need to find a more accurate verb. Perhaps something like "...informs Docker that the container will no longer listen on the specified..."

fredlf commented on docs/sources/reference/builder.md:
```diff
@@ -343,6 +346,27 @@ accessible from the host by default. To expose ports to the host, at runtime,
+    FROM debian:wheezy
+    EXPOSE 3000 8080
+
+We want to create a new image which exposes only port 3000. The
```
As per the style guide, "we" refers to Docker and "you" refers to the user. Please correct.

fredlf commented on docs/sources/reference/builder.md:
```diff
@@ -386,6 +410,28 @@ change them using `docker run --env <key>=<value>`.
+the list of available environment variables of the image and future build
+instructions.
+
+For example, given a image named `debug-app` defined with the following
```
an image

fredlf commented on docs/sources/reference/builder.md:
```diff
@@ -386,6 +410,28 @@ change them using `docker run --env <key>=<value>`.
+instructions.
+
+For example, given a image named `debug-app` defined with the following
+`Dockerfile`:
```
Incomplete sentence

fredlf commented on docs/sources/reference/builder.md:
```diff
@@ -386,6 +410,28 @@ change them using `docker run --env <key>=<value>`.
+    ENV DEBUG true
+
+We want to create a new image without the `DEBUG` environment variable. The
+`Dockerfile` for new image can be written as:
```
the new image

fredlf commented on docs/sources/reference/builder.md:
```diff
@@ -781,6 +827,33 @@ documentation.
+    FROM debian:wheezy
+    VOLUME ["/data", "/var/log"]
+
+we want to create a new image without the `/var/log` volume. The
```
Capitalize, correct pronoun.

fredlf commented:
Docs still need some corrections. Please refer to style guide here: http://docs.docker.com/contributing/docs_style-guide/

Thanks for the contribution!

dqminh commented:
@fredlf thanks for the review ! I pushed an update. PTAL.

fredlf commented on docs/sources/reference/builder.md:
```diff
@@ -343,6 +346,27 @@ accessible from the host by default. To expose ports to the host, at runtime,
+    EXPOSE 3000 8080
+
+You want to create a new image which exposes only port 3000. The
+`Dockerfile` for new image can be written as:
```
the new image

fredlf commented on docs/sources/reference/builder.md:
```diff
@@ -386,6 +410,28 @@ change them using `docker run --env <key>=<value>`.
+    ENV DEBUG true
+
+You want to create a new image without the `DEBUG` environment variable. The
+`Dockerfile` for the new image can be written as:
```
the new image (and below)

fredlf commented:
docs LGTM

erikh commented:
@dqminh can you get drone to pass? I'd like this in 1.6.

/cc @tiborvass

tiborvass commented:
This will require extensive testing

dqminh commented:
Sorry for the late response. I wasnt available for the past few days.

@erikh @tiborvass i rebased and the build passes, so i guess there were some problems with the old build. I think this is probably too late into 1.5 release cycle, so +1 for merging it for 1.6

dqminh commented:
@tiborvass i rebased again. Now that 1.5 is out (yay), lets try to get this into 1.6 :heart: . I dont want to be the one that has the oldest PR :older_man:

tiborvass commented:
*design review* (rebase is not necessary)

@dqminh @duglin @tianon 

Do you guys think `UNSET ENV a` should error out if there is no `ENV a` instruction in the dockerfile or any parent dockerfiles? I would vote for yes cause I like pedantic, but I am interested in your opinions.

duglin commented:
@tiborvass don't have time to read the entire thread, will try to do so tonight, but to your specific question... no I don't think it should error.  I view it like the bash "unset" command and that won't error if the variable isn't defined.

tianon commented:
@tiborvass I'm +1 on erroring out, especially since the same generic
"UNSET" instruction will be used for "VOLUME", etc, where an error would
definitely be warranted (so +1 for consistency)

thaJeztah commented:
In general I can be quite pedantic, but in this case I'm not sure; even if a `VOLUME` wasn't defined, wouldn't the end-result be the same? I see `UNSET` as an "idempotent" action; I want to make *sure* something isn't set, even if it wasn't set to begin with.

duglin commented:
@thaJeztah exactly.  Take the case where a base image sets an ENV but users of that image (child images) UNSET it.  Then one day the base image decided that it doesn't care about the ENV any more so it removes it.  Despite it having no negative impact on any child images, they'll all break on their next build because we generate an error on the UNSET.  

If we really want to give people the choice to be very very anal, then add a "-e" flag:
```
UNSET -e myvar
```
where -e means "error out if myvar isn't set".
But first we need https://github.com/docker/docker/pull/10775   :-)

tiborvass commented:
@thaJeztah @duglin but precisely, in your example, that would lead to `UNSET` pollution: all images 
Also, it's always easier to later remove the error if users really hate it, than to make a valid case erroneous later on.

thaJeztah commented:
> Also, it's always easier to later remove the error if users really hate it, than to make a valid case erroneous later on.

Okay, I'm with you on that. Can we come up with an "generic" error-message that's still useful enough to locate *which* "unset" caused the problem?

duglin commented:
@tiborvass I still go back to how the cmd line works. unset from sh doesn't generate an error and I think that's what people will naturally think of when they use this.

As for the example, I'm not inclined to think that breaking all possible users of an image is a good thing when the change has no real impact.  I'm not sure why UNSET pollution is an issue - if someone really wanted to they can use:  `UNSET --error-on-absent ....`  but I suspect most people won't use it.

dqminh commented:
I agree with @duglin that the unset instructions should not trigger an error that aborts the build if the property we want to unset does not exist in the original image, perhaps a warning is more suitable (though i'm not sure how to manipulate the build stdout to print out warning atm).

@tianon Can you explain how unset a non-existing volume in image definition will warrant an error ?

Also generic `UNSET` isntructions are no more, they are `UNSETENV | UNEXPOSE | NOVOLUME` now. But i guess the arguments are still applied.

duglin commented:
@dqminh the error you're seeing is because, for consistency, the error message should be of the form: ... requires at least one argument.  Unless you know exactly how many it needs, then its something like: ... requires exactly one argument.  See the other funcs for examples.

As for error or not... one more comment: I don't view UNSETENV as "undo a previous ENV" as much as "erase any value for env var xxx", or "make sure xxx has no value".  As a Dockerfile author when I use "UNSETENV" I'm doing so to make sure the env var in question is empty.  Whether it has a value "right now" or not is really irrelevant to me. I just want the net result to be that it is not set.

tiborvass commented:
Fine

@dqminh would you mind updating the original description to match the PR? You mentioned there is no more generic `UNSET` keyword.

tiborvass commented:
Ping @dqminh

dqminh commented:
@tiborvass I rebased and updated the original description.

duglin commented:
when done please squash your commits

dqminh committed:
add UNSETENV instruction to Dockerfile

UNSETENV can be used to remove unwanted environment variables from parent image

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

dqminh committed:
add UNEXPOSE instruction to Dockerfile

UNEXPOSE can be used to unset exposed ports from parent image

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

dqminh committed:
add NOVOLUME instruction to Dockerfile

NOVOLUME can be used to unset volumes defined in parent image.

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

dqminh committed:
add documentation for unsetenv, unexpose, novolume

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

duglin commented:
design LGTM

thaJeztah commented:
@dqminh another PR was just closed in favor of this one (https://github.com/docker/docker/issues/11921), however, that PR also included `LABEL`, which was added in docker 1.6.

Now that this is scheduled for docker 1.7, I think it makes sense to include `UNSETLABEL` as well?

dqminh commented:
@thaJeztah yah i think label will also be benefited from this. However I think it's better to focus on reviewing/merging the current changes. Adding label change will be straightforward after ( I hope ! )

thaJeztah commented:
@dqminh sounds fair. I thought it was worth mentioning to prevent it being forgotten :)

Hoping this will get merged soon!

LK4D4 commented:
@crosbymichael Would you mind to check UX? If it is okay, then we can proceed here.

tiborvass commented:
I'd love if we could check how this and nested builds play together / conflict. I start to think that nested builds would also solve this.

thaJeztah commented:
Sounds to me that nested builds are overkill for simple cases that people run into, for example;

```Dockerfile
FROM some:image
NOVOLUME /var/www
```

erikh commented:
I agree.

icecrime commented:
I'm pushing this to `2-needs-code-review`: it has been discussed at length, and ti seems that the design was already agreed long ago by Solomon.

tiborvass commented:
@icecrime @duglin @thaJeztah @erikh @tianon @bfirsh @proppy 

Sorry to be a PITA again. Started writing down my thoughts and then never hit the send button -_-
Not trying to bikeshed, just don't want to be in that bitter situation again of "oh it's already merged and we didn't think it through".

The problem nested builds solve is not the same as this PR, I'm aware of that. However, the main blocker on both PR, is syntax. And I believe that it just so happens they share the exact same syntax problem: you want one image (be it child image or nested image), to inherit only certain things from a parent image.

So, yes indeed, it is very tightly related to nested builds. It would be sad not to think of those two sharing a same syntax problem. NOVOLUME etc. is maybe good enough, but then it might be even more messy for nested builds. I'm not necessarily against the new instructions in this PR, I just want to make sure they are compatible with the nested builds idea, that's all.

**I tend to prefer white-lists to black-lists.** That said, there could be a usecase for black-lists too.

A rough example of what I'd prefer:
```
FROM image IMPORT {
    # only /some/path and /some/other/path would be imported
    FILE /some/path
    FILE /some/other/path
    # only `foo` env var would be imported, nothing else
    ENV foo
    # imports all volumes
    VOLUME
    # since there's no EXPOSE instruction, no exposed ports are imported
}
```

And for those who think nested builds is going to come in a long time, all I can say, is that users want it super badly, and I'll try my best to have it asap. The biggest issue is syntax, and that's not something we want to just do quickly.

erikh commented:
The thing I like about this approach is that it's fine-grained; I don't believe there's an overlap with nested builds on functionality, but perhaps there's room for both to solve similar problems in ways that meet specific needs.

I keep bringing this up, but f.e., I could see:

```
FROM foo
ENV bar quux
RUN do thing with ${bar}
NOENV bar
```

Which kind of capitalizes on the environment-as-dockerfile-variables pattern we have a lot of now, without leaving a bunch of garbage in the environment. I think @duglin may have fixed this specific problem however with changes specifically to how ENV works. This whole pattern could get more useful if we add conditionals (which last I checked, are still planned).

I'd love to pair with you on nested builds or this import code; I also want to add blocks to the parser but can do that independently.

That said, this should probably be run by solomon (or whoever makes these decisions these days) before we settle on anything. He already approved the design here, yet there is a lot of debate from others on how it should be carried out. Seems like a wise thing to do before merging anything at all.

-Erik

> On Apr 8, 2015, at 2:58 PM, Tibor Vass <notifications@github.com> wrote:
> 
> @icecrime <https://github.com/icecrime> @duglin <https://github.com/duglin> @thaJeztah <https://github.com/thaJeztah> @erikh <https://github.com/erikh> @tianon <https://github.com/tianon> @bfirsh <https://github.com/bfirsh> @proppy <https://github.com/proppy>
> Sorry to be a PITA again. Started writing down my thoughts and then never hit the send button -_-
> Not trying to bikeshed, just don't want to be in that bitter situation again of "oh it's already merged and we didn't think it through".
> 
> The problem nested builds solve is not the same as this PR, I'm aware of that. However, the main blocker on both PR, is syntax. And I believe that it just so happens they share the exact same syntax problem: you want one image (be it child image or nested image), to inherit only certain things from a parent image.
> 
> So, yes indeed, it is very tightly related to nested builds. It would be sad not to think of those two sharing a same syntax problem. NOVOLUME etc. is maybe good enough, but then it might be even more messy for nested builds. I'm not necessarily against the new instructions in this PR, I just want to make sure they are compatible with the nested builds idea, that's all.
> 
> I tend to prefer white-lists to black-lists. That said, there could be a usecase for black-lists too.
> 
> A rough example of what I'd prefer:
> 
> FROM image IMPORT {
>     # only /some/path and /some/other/path would be imported
>     FILE /some/path
>     FILE /some/other/path
>     # only `foo` env var would be imported, nothing else
>     ENV foo
>     # imports all volumes
>     VOLUME
>     # since there's no EXPOSE instruction, no exposed ports are imported
> }
> And for those who think nested builds is going to come in a long time, all I can say, is that users want it super badly, and I'll try my best to have it asap. The biggest issue is syntax, and that's not something we want to just do quickly.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/docker/docker/pull/8177#issuecomment-91048503>.
>

dqminh commented:
@erikh @tiborvass one benefit i see for this change is that it can also play nicely with current mechanism of `commit --change/import --change` when user wants to remove some properties of the parent's image. I think there should be use cases for both ( even if we add IMPORT whitelist properties ).

icecrime commented:
I understand @tiborvass wants consistency between image inheritance and the (future) nested builds, and that we should pick between a white-list (opt-in) or black-list (opt-out) approach now. My take on this is that the current behavior is "inherit everything", so I don't think we have a choice: it can only be black-list (opt-out) if we want to preserve backward-compatibility.

So in the end I'm +1 for this approach.

icecrime commented:
*Collective review with @LK4D4 @jfrazelle @crosbymichael @calavera*

I'm sorry, we don't understand valid use cases for this. Could you please bring us more use cases, or a link to a real world Dockerfile people use but have trouble inheriting from it because of the lack of such feature?

We're closing this as we mostly disagree, but please feel free to prove us wrong in the comments and we can reconsider!

thaJeztah commented:
Well, my main use case is still around volumes; there's no way to pre-populate a volume if the base image has a volume declaration (i.e., "package" an image for distribution). Using separate images for that works, but being able to "unset" a volume would have been a nice addition for some cases.

Ideally, but orthogonal to this, I'd like to see volumes not being used during build, but that has been discussed and is a BC breaking change (perhaps an `ONSTART` equivalent to `ONBUILD`)

I'll await progress on the nested builds and see how that looks

LK4D4 commented:
I think we should remove volumes from official images. With proper notifications and stuff.

duglin commented:
I'm not sure if this particular PR is the best solution for this but there was one situation that I thought was kind of good to support...  if a base image set an env var there's no way to unset it in a subsequent image.
For example:
```
echo -e FROM ubuntu\\nENV foo=bar | docker build -t test -
echo -e FROM test\\nENV foo=""\\nRUN env | docker build -
```
Anyone who uses 'test' image will always have `foo` set to something.
Now, perhaps we should look at each case individually, and in this case add an "unset" cmd.  IDK.

md5 commented:
> I think we should remove volumes from official images. With proper notifications and stuff.

@LK4D4 Really? You think that images like `mysql` should come with no volumes? I can't see how that's anything but a step backwards in terms of UX (not to mention performance).

I also don't understand why the discussion in #3465 (and possibly other tickets) has been consolidated into a PR that may be closed if this particular implementation is rejected.

md5 commented:
Also, how do you propose to properly notify users of official images? There's basically no mechanism to do so currently other than hoping that they've starred or otherwise followed the official image repo.

funkyfuture commented:
concerning ux, these both make sense:
- set reasonable volumes as @md5 points out
- unset these volumes in derived images

[that issue came up](https://github.com/samos123/docker-drupal/issues/10) on a Drupal-image that shall provide a simple vanilla-instance, but also be flexible enough for custom derivates.

though in that particular case the need to `UNVOLUME` arises from [this counter-intuitive and useless behaviour](https://github.com/docker/docker/issues/254#issuecomment-86689156) to provide a workaround.

@icecrime, imo these are real use-cases.

tianon commented:
I'm also definitely very much +1 on having a way to "UNVOLUME"; this is one
issue that comes up time and time again, and every time it does everyone's
standing around scratching their heads wondering why this isn't possible,
even though one of Docker's main strong points is the ability to re-use
other people's images instead of having to build your own.

Occasionally I also see the need for removing an environment variable, but
that one is much more rare than the need for removing defined volumes,
which crops up everywhere, even outside the official images.  I also
definitely agree with @md5 that removing the volume declarations completely
would be a large step backwards in terms of UX.

funkyfuture commented:
@icecrime @tianon as far as i understand the discussion there's a consent that 'unvolume' is a wanted feature. can you reopen this issue? or shall i open a new one?

gotcha commented:
We have the following use case: we want to build an image with postgres AND a dataset. 
This image is used by tests that, among others, update the data. By having the data inside the image, we ensure that changes remain confined in the container and that each test run in a new container can count on the same set of data.

Not having `UNVOLUME` or similar implies that we cannot use the current `postgres` official image as base image since it declares a volume.

ncronquist commented:
We have a use case for `UNSETENV`. During our build process, we set some credential environment variables to allow us to install private node modules inside the container. The installation of those modules needs to be inside the container because some of the modules require OS specific binaries. After we build the container, we would like to remove the credential environment variables. Currently we just set them to nothing, but being able to unset them would be preferred. This seems like a problem that other teams have probably experienced, so maybe there is a better way to accomplish what we're trying to do.

thaJeztah commented:
@ncronquist docker 1.9 will have build-time variables, but (just like `ENV`) we discourage using them for passing secrets. You can use them to customize builds, though: see https://github.com/docker/docker/issues/14634

l3iggs commented:
+1 for UNSETENV

sporkmonger commented:
+1 for UNVOLUME

We're deploying read-only databases in some cases. Being able to treat the image as a self-contained unit that avoids the complexity of volume management in our cluster infrastructure when it's not required is valuable (by baking the immutable data into the child image). However, being able to inherit from a database base image that isn't concerned with the read-only use-case is equally valuable.

modius commented:
+1 UNEXPOSE

Apologies for the repost.. this looks like a more appropriate place:

> There are several solutions that read the container metadata and provide additional upstream configuration. For example; with HAPROXY i have to set EXCLUDE_PORTS=8080 to block it trying to dynamically provide access to that port on my Tomcat apps.
> 
> Developers look at the exposed ports and make assumptions about the behaviour of the container. For example I have a base image that extends Tomcat (EXPOSE 8080) but the image uses a different port to the default (EXPOSE 8888). If you add a web server in a compound image (eg. running NGINX and Tomcat) you serve content via HTTP (EXPOSE 80 and EXPOSE 443).
> 
> In the latter example you can end up with an image that self documents as exposing 8080, 8888, 80 and 443 where only 80/443 is relevant.
> 
> These are real problems as evidenced by the fact I have to keep explaining things to developers in our community despite very specific documentation; who needs documentation when you can just look at the image? <-- everyone when the image is self-documenting the wrong thing.
> 
> There are workarounds for this problem but they are workarounds. Is this a major architectural issue? Why can't Docker consider a more elegant solution to this real problem.

https://github.com/docker/docker/issues/3465
https://github.com/lucee/lucee-dockerfiles/issues/13

cmehay commented:
I have an usecase for UNEXPOSE

Linking a container with exposed port to another one results some environment variables to be created on the container with information about exposed ports and environment.

I'm using these variables with a tool to manage auto-configuration in containers, if there is an exposed port which is not actually used, my configuration will be corrupted :disappointed: 

There is no way to reset exposed port from a container, you can only add some but nothing to close one.

So, I think that UNEXPOSE is an important feature.

juliangamble commented:
Use case for UNEXPOSE

Suppose I have four docker hosts and I want to run 16 maven tomcat containers which all default to internal port 8080.

Now imagine I'm using registrator with rancher CNI - which locks me down to the internal port. 
https://github.com/gliderlabs/registrator/issues/541#issuecomment-305012416
This means I can only run one 8080 internal port per host. (Since I have to do 8080:8080 port mappings)

In this situation - docker's internal->external port mapping isn't enough to solve my problem. I actually need to override the internal port mapping, preferably without rebuilding the original container.

maxgorovenko commented:
You guys spent almost 4 years for talking about UI and command name? Really?

jimethn commented:
As a programmer, I had more emotional highs and lows reading through this than I'll ever get from a TV show. Came here looking for a way to UNVOLUME in 2017. Left disappointed.

edevil commented:
Just today I needed to run a wordpress site and I faced the same problem, it declares a ```/var/www/html``` volume: https://github.com/docker-library/wordpress/blob/master/Dockerfile-debian.template#L28

I wanted to include the content via a child image but could not...

ibc commented:
Excuse me, is this implemented or not?

thaJeztah commented:
@ibc no it wasn't; see https://github.com/moby/moby/issues/3465#issuecomment-356988520

juliangamble commented:
Thanks @thaJeztah - for us to do a pull request - could you please point to the line of code (in your point of view) for us to start reading.

gdraheim commented:
As in the dicussion of #3465 please note that there is a workaround available (using docker-copyedit).

kinglion811 commented:
the UNSETENV or UNSET ENV can work now,how can I delete some base image's env when I build image

chpock commented:
+UNVOLUME

Athaphian commented:
Why was this closed due to no real world cases? This happens pretty often right?

We build a bunch of docker images that use each other as base image. In one of the base images we set ENV CYPRESS_BINARY_VERSION because all our projects require this. But now we want to upgrade Cypress to a new version for one project at the time and this version requires CYPRESS_BINARY_VERSION to not exist... so we need to override this in one of the project docker images but not change the base image.. only UNSET does not exist! This was started in 2014, it could have been fixed in 2017, but still almost 2019 and not fixed?

Athaphian commented:
@gdraheim Im sure that will work, and we might eventually go that route too. But it is more tedious than just adding unsetenv in the dockerfile.

akramfares commented:
+UNVOLUME

syedhassaanahmed commented:
Really need an UNVOLUME option. docker-copyedit is a clunky option.

ebuildy commented:
Sorry to "exhume" this wonderful PR, but clearly remove ``env`` / ``volume`` / ``port`` from base image make sense in order to avoid bad surprise (such as a ton of abandoned volumes, unwanted exposed ports ...), the PR seem well documented and tested, why this is not merged yet?

Or we should strongly discourage to use such properties in base image? (I am in trouble with the official image https://github.com/docker-library/docker/blob/65fab2cd767c10f22ee66afa919eda80dbdc8872/18.09/dind/Dockerfile#L40 that create a volume where Docker images are stored, in CI env, this is really unwanted).

Thanks you

seraasch commented:
Also have a use case for UNEXPOSE.

Why is this marked closed?

markmsmith commented:
@icecrime Sorry to `@` you, but I was just curious if there were any plans to revisit this PR, now that several real world use cases have been collected above.  It looks like https://github.com/moby/moby/issues/3465 has been reopened, so does that mean that this is back on the table?  

I ran in to this problem today using the official Postgres image with the TestContainers project (https://github.com/testcontainers/testcontainers-node), where my integration tests started generating volumes from the containers.  It appears that the `docker run --rm` command has been enhanced to remove unnamed volumes (https://github.com/moby/moby/pull/19568), but this doesn't help when working with the container programatically.  
I'm hoping that there will be a solution in the future to allow extending a base image and unset a volume from the parent.

zengqingfu1442 commented:
@dqminh Thanks, are these commands available now in Dockerfile?

> New added instructions are:
>     UNSETENV ENV1 ENV2 ... unset environment variables inherited from parent image
>    UNEXPOSE PORT ... remove exposed ports inherited from parent image
>    NOVOLUME PATH remove volume mountpoints from parent image

yywing commented:
+UNVOLUME

simaotwx commented:
I'm all for UNVOLUME which would be really helpful for a number of scenarios where all someone wants is to remove a volume from a base image and nothing else.

alexeyp0708 commented:
@icecrime 
> _Collective review with @LK4D4 @jfrazelle @crosbymichael @calavera_
> 
> I'm sorry, we don't understand valid use cases for this. Could you please bring us more use cases, or a link to a real world Dockerfile people use but have trouble inheriting from it because of the lack of such feature?
> 
> We're closing this as we mostly disagree, but please feel free to prove us wrong in the comments and we can reconsider!



> [icecrime](/icecrime)

Arguments in favor added UNVOLUME.
The problem exists and people suffer from it and ask for something to be done.
1. The child image does not always need the volumes of the parent image.
2.  There is such a problem. https://github.com/moby/moby/issues/33842#issuecomment-1776771402
3.  When changing daemon settings, developers sometimes prefer to specify their own volumes for worker processes.
As a result, parent volumes hang like extra garbage by default.
4.  The developer would prefer to have access to the folder in the container that the volume overlaps.
But the default volume always overlaps this folder in the container and the user does not have access to it.

Please study the topic https://github.com/moby/moby/issues/3465 and you will understand that there are problems. It’s just that not everyone can or wants to describe the situation that the developers encountered.

simaotwx commented:
> Collective review with @LK4D4 https://github.com/jfrazelle @crosbymichael @calavera
> 
> I'm sorry, we don't understand valid use cases for this. Could you please bring us more use cases, or a link to a real world Dockerfile people use but have trouble inheriting from it because of the lack of such feature?
> 
> We're closing this as we mostly disagree, but please feel free to prove us wrong in the comments and we can reconsider!

My use case is https://github.com/docker-library/wordpress/blob/master/latest/php8.1/fpm/Dockerfile#L151 where I want to add a customized wordpress install directly to a downstream image but I have to resort to copying the **entire** file, just to remove that `VOLUME`.

I saw that you can use `COPY --from base / /` but I'm not sure if that would introduce edge cases and weird behavior with symlinks, xattrs, ACLs, permissions, ownership, creation/modification time and special files.

thaJeztah commented:
Sorry, but I'm locking the conversation on this PR; this PR is 10 years old, including decisions made 10 Years ago.

The Dockerfile syntax and Builder is no longer maintainer in this repository, and now maintained by the BuildKit project (https://github.com/moby/buildkit), which is where features like this can be implemented.

While this original PR was rejected, the ticked related to this PR was reopened, but the overall design may need discussion. Also note that some of the use-cases (`UNVOLUME` or `UNSET VOLUME` to prevent volumes from being applied during build) changed with BuildKit, which does not apply volumes _during build_.

This comment https://github.com/moby/moby/issues/3465#issuecomment-356988520 is relevant here;

> @veqryn since reopening this issue, nobody started working on a pull-request; the existing pull request did no longer apply cleanly on the code-base so a new one has to be opened; if anyone is interested in working on this, then things can get going again.
>
> See my earlier comment(s); [#3465 (comment)](https://github.com/moby/moby/issues/3465#issuecomment-247405438)
>
> > We discussed this issue in the maintainers meeting, and generally, we're OK to start working on this again.
> > @duglin are you perhaps interested in working on this?
>
> And [#3465 (comment)](https://github.com/moby/moby/issues/3465#issuecomment-313549657)
>
> > As to why it isn't there yet; simply because nobody has had time to start working on it, but if someone is interested, it most likely will be accepted.


These things are best discussed in the BuildKit issue tracker, in which a design can be proposed.
